sktree/stats/_might.py

```diff
@@ -0,0 +1,317 @@
+            )
+        elif stat == "MI":
+            H_YX = np.mean(entropy(posterior_final[:, 1], base=np.exp(1)))
+            _, counts = np.unique(posterior_final[:, 0], return_counts=True)
+            H_Y = entropy(counts, base=np.exp(1))
+            self.stat = max(H_Y - H_YX, 0)
+
+        if return_pos:
+            return self.stat, posterior
+
+        return self.stat
+
+    def test(self, x, y, reps=1000, workers=1, random_state=None):
+        observe_stat = self.statistic(x, y)
+
+        null_dist = np.array(
+            Parallel(n_jobs=workers)([delayed(perm_stat)(self, x, None, y) for _ in range(reps)])
+        )
+        pval = (1 + (null_dist >= observe_stat).sum()) / (1 + reps)
+
+        return observe_stat, null_dist, pval
+
+
+class MIGHT_MV:
+    def __init__(
+        self,
+        n_estimators=500,
+        criterion="gini",
+        splitter="best",
+        max_depth=None,
+        min_samples_split=2,
+        min_samples_leaf=1,
+        min_weight_fraction_leaf=0.0,
+        max_features="sqrt",
+        max_leaf_nodes=None,
+        min_impurity_decrease=0.0,
+        bootstrap=False,
+        oob_score=False,
+        n_jobs=None,
+        random_state=None,
+        verbose=0,
+        warm_start=False,
+        class_weight=None,
+        ccp_alpha=0.0,
+        max_samples=None,
+        honest_prior="empirical",
+        honest_fraction=0.5,
+        tree_estimator=None,
+        limit=0.05,
+    ):
+        self.clf = HonestForestClassifier(
+            n_estimators=n_estimators,
+            criterion=criterion,
+            splitter=splitter,
+            max_depth=max_depth,
+            min_samples_split=min_samples_split,
+            min_samples_leaf=min_samples_leaf,
+            min_weight_fraction_leaf=min_weight_fraction_leaf,
+            max_features=max_features,
+            max_leaf_nodes=max_leaf_nodes,
+            min_impurity_decrease=min_impurity_decrease,
+            bootstrap=bootstrap,
+            oob_score=oob_score,
+            n_jobs=n_jobs,
+            random_state=random_state,
+            verbose=verbose,
+            warm_start=warm_start,
+            class_weight=class_weight,
+            ccp_alpha=ccp_alpha,
+            max_samples=max_samples,
+            honest_prior=honest_prior,
+            honest_fraction=honest_fraction,
+            tree_estimator=tree_estimator,
+        )
+        self.limit = limit
+
+    def statistic(
+        self,
+        x,
+        y,
+        stat="AUC",
+        workers=1,
+        test_size=0.2,
+        initial=True,
+        return_pos=False,
+        permute_y=False,
+    ):
+        # Initialize trees
+        if initial:
+            self.clf.fit(x[0:2], y.ravel()[0:2])
+
+        # Compute posteriors with train test splits
+        posterior = Parallel(n_jobs=workers)(
+            delayed(auc_calibrator)(tree, x, y.ravel(), test_size, permute_y)
```

Comment:
But this then does not work for the problem that the Coleman paper addresses right?

If we are interested in whether or not X1 is important for predicting Y in terms of AUC, for X = [X1, X2, ..., Xd], then permuting Y destroys the dependency of Y on all other X?

Or perhaps this design was to address the problem of whether or not generally X (the whole column space) provides information on Y? If so, then sure this is valid, but if we permute the row indices of X, then it can cover both use-cases.

cc: @jovo

Reply:
I assume this is only for null observations, but yeah on multi-view it would not make sense.

Reply:
the key is for each tree to do the permutation independently, Coleman only did 1 permutation.  i do not feel strongly about whether to permute X or Y, for our test, it doesn't matter.